File: src/main/java/com/study/platform/global/response/ApiResponse.java
package com.study.platform.global.response;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.study.platform.global.exception.ErrorCode;
import lombok.Getter;
import org.springframework.http.ResponseEntity;

@Getter
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ApiResponse<T> {

    private final boolean success;
    private final String code;
    private final String message;
    private final T data;

    private ApiResponse(boolean success, String code, String message, T data) {
        this.success = success;
        this.code = code;
        this.message = message;
        this.data = data;
    }

    public static <T>ResponseEntity<ApiResponse<T>> success(SuccessCode successCode, T data) {
        return ResponseEntity
                .status(successCode.getHttpStatus())
                .body(new ApiResponse<>(true, successCode.name(), successCode.getMessage(), data));
    }

    public static ResponseEntity<ApiResponse<Void>> success(SuccessCode successCode) {
        return ResponseEntity
                .status(successCode.getHttpStatus())
                .body(new ApiResponse<>(true, successCode.name(), successCode.getMessage(), null));
    }

    public static ResponseEntity<ApiResponse<Void>> fail(ErrorCode errorCode) {
        return ResponseEntity
                .status(errorCode.getHttpStatus())
                .body(new ApiResponse<>(false, errorCode.name(), errorCode.getMessage(), null));
    }
}

File: common/src/main/kotlin/com/study/platform/global/response/ApiResponse.kt
package com.study.platform.global.response

import com.fasterxml.jackson.annotation.JsonInclude
import com.study.platform.global.exception.ErrorCode
import org.springframework.http.ResponseEntity

@JsonInclude(JsonInclude.Include.NON_NULL)
class ApiResponse<T> private constructor(
    val success: Boolean,
    val code: String,
    val message: String,
    val data: T?
) {
    companion object {
        fun <T> success(successCode: SuccessCode, data: T): ResponseEntity<ApiResponse<T>> =
            ResponseEntity
                .status(successCode.httpStatus)
                .body(ApiResponse(true, successCode.name, successCode.message, data))

        fun success(successCode: SuccessCode): ResponseEntity<ApiResponse<Void>> =
            ResponseEntity
                .status(successCode.httpStatus)
                .body(ApiResponse(true, successCode.name, successCode.message, null))

        fun fail(errorCode: ErrorCode): ResponseEntity<ApiResponse<Void>> =
            ResponseEntity
                .status(errorCode.httpStatus)
                .body(ApiResponse(false, errorCode.name, errorCode.message, null))

        fun noContent(): ResponseEntity<Void> = ResponseEntity.noContent().build()
    }
}
